### PR TITLE
Extended Attributes: Use LaunchServices API for quarantine data 

### PR DESCRIPTION
**Pull request by sharvilshah**
Now uses LaunchServices API to grab quarantine data.

* No more manual parsing (often prone to breaking on updates)
* More insight into quarantine events: timestamp, type (webdownload, mail.app attachment, ical attachment, airdrop), etc 
* Added tests

```
osquery> select key,value,base64 from extended_attributes where path = "/Users/sbs/Downloads/foo_bar/googlechrome.dmg";
+----------------------------------+------------------------------------------------------------------+--------+
| key                              | value                                                            | base64 |
+----------------------------------+------------------------------------------------------------------+--------+
| com.apple.diskimages.fsck        | OP++fu++J8B7AGyIsqFQTRZwbHo=                                     | 1      |
| com.apple.diskimages.recentcksum | (truncated)                                                      | 0      |
| where_from                       | https://dl.google.com/chrome/mac/stable/GGRO/googlechrome.dmg    | 0      |
| where_from                       | https://www.google.com/chrome/browser/thankyou.html?platform=mac | 0      |
| quarantine_agent                 | Google Chrome                                                    | 0      |
| quarantine_data_url              | https://dl.google.com/chrome/mac/stable/GGRO/googlechrome.dmg    | 0      |
| quarantine_event_id              | 6A0EC92D-1882-4358-A315-586BA65F8F77                             | 0      |
| quarantine_origin_url            | https://www.google.com/chrome/browser/thankyou.html?platform=mac | 0      |
| quarantine_timestamp             | 1430867421                                                       | 0      |
| quarantine_type                  | LSQuarantineTypeWebDownload                                      | 0      |
+----------------------------------+------------------------------------------------------------------+--------+
```

```
osquery> select key,value,base64 from extended_attributes where path = "/Users/sbs/Downloads/foo_bar/foobar.jpg";
+-----------------------------------------+---------------------------------------+--------+
| key                                     | value                                 | base64 |
+-----------------------------------------+---------------------------------------+--------+
| where_from                              | iPhone                                | 0      |
| com.apple.metadata:kMDLabel_(truncated) | (truncated)                           | 1      |
| quarantine_agent                        | sharingd                              | 0      |
| quarantine_event_id                     | 5E928A03-6E22-4B5D-A5D8-F9C19F23ED9A  | 0      |
| quarantine_sender                       | iPhone                                | 0      |
| quarantine_timestamp                    | 1430940342                            | 0      |
| quarantine_type                         | LSQuarantineTypeAirDrop               | 0      |
+-----------------------------------------+---------------------------------------+--------+
```

**profiler:**

```bash
❯ sudo -E python ./tools/profile.py --leaks --query "select * from extended_attributes where directory = \"/Users/sbs/Downloads/foo_bar\";"
Analyzing leaks in query: select * from extended_attributes where directory = "/Users/sbs/Downloads/foo_bar";
  definitely: 0 leaks for 0 total leaked bytes.

❯ sudo -E python ./tools/profile.py --leaks --query "select * from extended_attributes where path = \"/Users/sbs/Downloads/foo_bar/foobar.jpg\";"
Analyzing leaks in query: select * from extended_attributes where path = "/Users/sbs/Downloads/foo_bar/foobar.jpg";
  definitely: 0 leaks for 0 total leaked bytes.

❯ sudo -E python ./tools/profile.py --leaks --query "select * from extended_attributes where path = \"/Users/sbs/Downloads/foo_bar/googlechrome.dmg\";"
Analyzing leaks in query: select * from extended_attributes where path = "/Users/sbs/Downloads/foo_bar/googlechrome.dmg";
  definitely: 0 leaks for 0 total leaked bytes.
```

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by theopolis** [comment]
ok to test

File: osquery/tables/system/darwin/extended_attributes.cpp [hunk 1]
```diff
@@ -164,72 +130,87 @@ void parseWhereFrom(QueryData &results,
...
+                     kCFAbsoluteTimeIntervalSince1970;
+    value = INTEGER(std::llround(unix_time));
+  } else if (CFGetTypeID(property) == CFURLGetTypeID()) {
+    value = stringFromCFString(CFURLGetString((CFURLRef)property));
+  }
+  setRow(results, path, table_key_name, value);
+}
+
+void parseQuarantineFile(QueryData &results, const std::string &path) {
+  CFURLRef url = CFURLCreateFromFileSystemRepresentation(
```
Comment (theopolis) [inline osquery/tables/system/darwin/extended_attributes.cpp]: Could this ever return a `nullptr`?

File: osquery/tables/system/darwin/extended_attributes.cpp [hunk 2]
```diff
@@ -164,72 +130,87 @@ void parseWhereFrom(QueryData &results,
...
+    CFStringRef key = CFStringCreateWithCString(
+        kCFAllocatorDefault, kv.second.c_str(), kCFStringEncodingUTF8);
+
+    if (CFDictionaryGetValueIfPresent((CFDictionaryRef)quarantine_properties,
+                                      key, &property)) {
+      extractQuarantineProperty(kv.first, property, path, results);
+    }
+    CFRelease(key);
+  }
+  CFRelease(property);
```
Comment (theopolis) [inline osquery/tables/system/darwin/extended_attributes.cpp]: I don't think you need to release `property`.

File: osquery/tables/system/darwin/extended_attributes.cpp [hunk 3]
```diff
@@ -164,72 +130,87 @@ void parseWhereFrom(QueryData &results,
...
+    return;
+  }
+#else
+  if (!CFURLCopyResourcePropertyForKey(url, kCFURLQuarantinePropertiesKey,
+                                       &quarantine_properties, nullptr)) {
+    VLOG(1) << "Error retrieving quarantine properties for " << path;
+    CFRelease(url);
     return;
   }
+#endif
```
Comment (theopolis) [inline osquery/tables/system/darwin/extended_attributes.cpp]: Is it possible for `LSCopyItemAttribute` or `CFURLCopyResourcePropertyForKey` to return indicating success and not have set `quarantine_properties`. aka, is it worth checking `quarantine_properties != nullptr` here?

File: osquery/tables/system/darwin/extended_attributes.cpp (hunk 1 above)
Comment (sharvilshah) [inline osquery/tables/system/darwin/extended_attributes.cpp]: Yup, missed that one while moving code around. Good catch.
From Apple's `CFURL` docs:
> If a function fails to create a CFURL object, it returns NULL, which you must be prepared to handle.

File: osquery/tables/system/darwin/extended_attributes.cpp (hunk 3 above)
Comment (sharvilshah) [inline osquery/tables/system/darwin/extended_attributes.cpp]: As far as I can tell I don't think so, but still can be worthwhile to add that check and break early.
On `CFURLCopyResourcePropertyForKey`: 

> Return Value
>        true if propertyValueTypeRefPtr is successfully populated; false if an error occurs.

And `LSCopyItemAttribute` returns `OSStatus` and I am checking it it's `noErr`

**Comment by sharvilshah** [comment]
Thanks for the quick review @theopolis. Updated.

Ran profile.py again, no change.

**Comment by theopolis** [comment]
Awesome! If you want to be super duper doubly sure, remove your `./build` directory and run `make analyze`. It'll take some time, but it will swap your compiler with a script that runs clang/LLVM's static analyzer. I run this before each release cut so do not worry about having it as part of a regular build/test iteration.

**Pull request merged**